consensus/hotstuff/committees/consensus_committee_test.go

```diff
@@ -230,32 +208,85 @@ func (suite *ConsensusSuite) TestProtocolEvents_CommittedEpoch() {
+
+	suite.committee.EpochExtended(suite.currentEpochCounter, refBlock, extension)
 	// wait for the protocol event to be processed (async)
 	require.Eventually(suite.T(), func() bool {
 		_, err := suite.committee.IdentitiesByEpoch(unittest.Uint64InRange(201, 300))
 		return err == nil
-	}, 30*time.Second, 50*time.Millisecond)
+	}, time.Second, 50*time.Millisecond)
 
-	suite.Assert().Len(suite.committee.epochs, 2)
-	suite.AssertStoredEpochCounterRange(suite.currentEpochCounter, suite.currentEpochCounter+1)
+	// we should have the same number of cached epochs (an existing epoch has been extended
+	suite.Assert().Len(suite.committee.epochs, 1)
+	suite.AssertStoredEpochCounterRange(suite.currentEpochCounter, suite.currentEpochCounter)
 
 	// should handle multiple deliveries of the protocol event
-	suite.committee.EpochFallbackModeTriggered(0, nil)
-	suite.committee.EpochFallbackModeTriggered(0, nil)
-	suite.committee.EpochFallbackModeTriggered(0, nil)
+	suite.committee.EpochExtended(suite.currentEpochCounter, refBlock, extension)
+	suite.committee.EpochExtended(suite.currentEpochCounter, refBlock, extension)
+	suite.committee.EpochExtended(suite.currentEpochCounter, refBlock, extension)
 
-	suite.Assert().Len(suite.committee.epochs, 2)
-	suite.AssertStoredEpochCounterRange(suite.currentEpochCounter, suite.currentEpochCounter+1)
+	suite.Assert().Len(suite.committee.epochs, 1)
+	suite.AssertStoredEpochCounterRange(suite.currentEpochCounter, suite.currentEpochCounter)
+	_, err := suite.committee.IdentitiesByEpoch(unittest.Uint64InRange(201, 300))
+	suite.Assert().NoError(err)
+}
+
+// TestProtocolEvents_EpochExtendedMultiple tests that protocol events notifying of an epoch extension are handled correctly.
+// An EpochExtension event should result in a re-computation of the leader selection (including the new extension).
+// The Committee should handle multiple subsequent, contiguous epoch extensions.
+// Repeated events should be no-ops.
+func (suite *ConsensusSuite) TestProtocolEvents_EpochExtendedMultiple() {
+	curEpoch := newMockEpoch(suite.currentEpochCounter, unittest.IdentityListFixture(10), 101, 200, true)
+	suite.epochs.Add(curEpoch)
+
+	suite.CreateAndStartCommittee()
+
+	// Add several
+	for i := 0; i < 10; i++ {
+		finalView, err := curEpoch.FinalView()
+		require.NoError(suite.T(), err)
+		extension := flow.EpochExtension{
+			FirstView: finalView + 1,
+			FinalView: finalView + 100,
+		}
+		refBlock := unittest.BlockHeaderFixture()
+		addExtension(curEpoch, extension)
+		suite.state.On("AtBlockID", refBlock.ID()).Return(suite.snapshot)
+
+		suite.committee.EpochExtended(suite.currentEpochCounter, refBlock, extension)
+		// wait for the protocol event to be processed (async)
+		require.Eventually(suite.T(), func() bool {
+			_, err := suite.committee.IdentitiesByEpoch(unittest.Uint64InRange(extension.FirstView, extension.FinalView))
+			return err == nil
+		}, time.Second, 50*time.Millisecond)
+
```

Comment:
Since this is primarily to wait for a condition to occur, I've replaced this and similar checks within `Eventually` blocks with queries for the first view of the next extension in all cases. Subsequently we check the larger range of expected views, but we don't need to do it twice.